### ISSUE TITLE
Clean up x, y vs row, col vs width, height in C++ code

### ISSUE THREAD

**Issue by cjh1** [OpenChemistry/stempy#108]
While we are doing #106, I would like to do a similar pass through the C++ code. I can see at least one place where there may be a discrepancy:

https://github.com/OpenChemistry/stempy/blob/master/stempy/image.h#L53

**Comment by cjh1** [comment]
I think we should use `std::pair` to represent dimensions and x, y coordinates ( I think then pybind11 will take care of the convertions for us). We can create a few type definitions for example:

`using Dimensions = std::pair<int, int>; // width, height`
`using Coordinates = std::pair<int, int>;

or alternatively we could create a new simple `struct`'s, but I'm not sure how we would get pybind11 to do the correct convertion from `tuple`.

x, y = width, height. Lets replace any uses of rows and columns.

**Comment by alesgenova** [comment]
Unless I'm getting confused, I think the convention in numpy is the following though:

```python
shape = (n_rows, n_cols)  # i.e. (height, width)
```

**Comment by cjh1** [comment]
@alesgenova This is true but I guess in general when talking about images or 2d graphic x is the width and y is the height ...

**Comment by psavery** [comment]
@cjh1 For that example you mentioned [here](https://github.com/OpenChemistry/stempy/blob/master/stempy/image.h#L53), it looks like those two variables are mistakenly labeled `rows` and `columns` in the header file, when in the CPP file and in all of the use cases of it (in python included), they are treated as `width` and `height`. So it looks to me that it is just a labeling problem in the header, and that gets fixed in #109.

**Comment by psavery** [comment]
However, I did end up switching [these two variables](https://github.com/OpenChemistry/stempy/pull/109/files#diff-5afbef7ae6ecfa409c949d37ebfd1e93R90) in #109, as they seemed out of order. We'll probably need to do some tests on some non-square data to confirm.

**Comment by ercius** [comment]
@psavery @cjh1 I added an example data set which is non-square and small (11GB). See `ncemhub/4Dcamera/2020.01.13/data_scan137*.data`. You can count this in a few seconds (background_threshold_n_sigma = 4; background_threshold = 28.62) and create stem images very quickly too. If you get the scan width and height incorrect it's obvious to see the mistake in a stem image. The original data was acquire with 129 columns (x, width) and 128 rows (y, height) on the microscope. So, stempy should try to match this.

The correct ADF was created using this python command:
```python
ims = stim.create_stem_images_sparse(ee,(0,200),
(100,288),width=128,height=129,frame_height=576,frame_width=576)
```

The correct ADF image looks like this:
![scan137_adf](https://user-images.githubusercontent.com/15657108/72922073-3f086d80-3d01-11ea-8f6d-3dac350be3d8.jpg)

Flipping the height and width creates an incorrect image:
```python
ims = stim.create_stem_images_sparse(ee,(0,200),
(100,288),width=129,height=128,frame_height=576,frame_width=576)
```
The incorrect ADF image looks like this:
![scan137_adf_flipped](https://user-images.githubusercontent.com/15657108/72922093-46c81200-3d01-11ea-8d58-8c8e39293d5b.jpg)

Further, these images are rotated 90 degrees from what the scanning looks like in the microscope. You can tell this (subtly) by looking at the left hand side of the correct image above. There is some lost data in the first few scan lines (a known issue) which should happen in the first few rows (not columns). So, there is a 90 rotation clock-wise that is needed to make stempy match what the user sees on the microscope. Not sure if this is a write out issue (I used PIL.Image) or a stempy issue. I will acquire a more obvious non-square image (e.g. 129x512) an upload that to further test this out.

Note: All of the data we acquire is non-square due to a "fly-back" column which should be ignored in all data sets. So, the column number will always be +1 off from what we actually intended ( a data set will be 128x129 where 129 is 128 data  + 1 fly-back column).

**Comment by psavery** [comment]
The swapped width and height might be potentially fixed by [this change](https://github.com/OpenChemistry/stempy/pull/109/files#diff-5afbef7ae6ecfa409c949d37ebfd1e93R90) that is part of #109 (the width and height were swapped there). But we should test it out to be sure.

**Comment by cjh1** [comment]
Thanks @erius, we will test this out. @psavery I will download the dataset so I can give you access to try it out.

**Comment by psavery** [comment]
Unfortunately, my changes don't appear to have fixed that problem.

@ercius We read in the data as a 1D array and keep it in a one-dimensional form for most of the operations. Unless I'm mistaken, it looks like the only place where the scan dimension order is used (and actually matters) is when pybind11 converts it to a numpy array [here](https://github.com/OpenChemistry/stempy/blob/4b99578befea22d2dbc9f557c40aeab4eba4b982/python/image.cpp#L54-L64). And it looks like we are ordering it according to their instructions.

Do you know if the data is being written out in column-major (Fortran) or row-major (C) ordering?

I wonder if the fact that the correct result is rotated by 90 degrees is why the width and height are switched...

**Comment by psavery** [comment]
@ercius Nevermind. I think we figured it out. Two positions were swapped in our header reader...

**Comment by ercius** [comment]
@psavery Great! Let me know when everything is pulled into master so I can update our analysis PC?

**Comment by cjh1** [comment]
@ercius We will try to get this merged today.

**Comment by psavery** [comment]
@ercius The changes are not quite merged yet, but here is what a stem image generated from the raw data with scan dimensions (129, 128) looks like:
![129_128_stem_image](https://user-images.githubusercontent.com/9558430/73379136-ba34cb00-428f-11ea-95d4-9894146ca57a.png)

Same thing except generated from sparse data:
![129_128_sparse_stem_image](https://user-images.githubusercontent.com/9558430/73379228-e3555b80-428f-11ea-8fd5-75079aca66f9.png)

Also, you will now be able to get the same results without specifying the scan dimensions, because now it correctly reads them from the header file as well.

Let me know if the images look good.

**Comment by cjh1** [comment]
@ercius All merges, let us know if you run into any issues. I am rebuild the shifter image on cori as well ( should be done in a couple of minutes ).

**Comment by ercius** [comment]
stempy is not working with the current master. I get an error for electron_count when I pass in a sector reader and a np.zeros((576,576)) as the dark. Its says that the inputs are incorrect:
`stim.electron_count(reader,np.zeros((576,576)))`

Here is the error:
```
TypeError: electron_count(): incompatible function arguments. The following argument types are supported:
    1. (arg0: stempy::BlockIterator<stempy::StreamReader>, arg1: stempy::BlockIterator<stempy::StreamReader>, arg2: stempy._image._image_double, arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    2. (arg0: stempy::BlockIterator<stempy::SectorStreamReader>, arg1: stempy::BlockIterator<stempy::SectorStreamReader>, arg2: stempy._image._image_double, arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    3. (arg0: stempy::PyReader::iterator, arg1: stempy::PyReader::iterator, arg2: stempy._image._image_double, arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    4. (arg0: stempy::BlockIterator<stempy::StreamReader>, arg1: stempy::BlockIterator<stempy::StreamReader>, arg2: numpy.ndarray[float64], arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    5. (arg0: stempy::BlockIterator<stempy::SectorStreamReader>, arg1: stempy::BlockIterator<stempy::SectorStreamReader>, arg2: numpy.ndarray[float64], arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    6. (arg0: stempy::PyReader::iterator, arg1: stempy::PyReader::iterator, arg2: numpy.ndarray[float64], arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data

Invoked with: <stempy._io._sector_reader_iterator object at 0x7f85cf31a848>, <stempy._io._sector_reader_iterator object at 0x7f85cf3325a8>

**Comment by cjh1** [comment]
@ercius @psavery Is taking a look at this.

**Comment by cjh1** [comment]
@psavery I am guessing that one of the template instantiations is missing?

**Comment by psavery** [comment]
@ercius I haven't been able to reproduce it so far. I'm calling it like this:

```
# The scan_dimensions here are in the correct order, but the image is wrong
reader = io.reader(files, version=io.FileVersion.VERSION4)
electron_counted_data = image.electron_count(reader, np.zeros((576, 576)))
```

Are you passing `io.FileVersion.VERSION4` to the reader?

**Comment by psavery** [comment]
Or maybe you can make sure you did a `pip install -e .`, to ensure that the python is up-to-date?

**Comment by cjh1** [comment]
It would actually be `pip install -U -e .`

**Comment by ercius** [comment]
The issue with pip was the first thing I tried. I just tried it again and I got a slightly different version number. In my history for pip install I see `stempy==0.1.dev349`. After running it again I see `stempy==0.1.dev404'. Im not sure whats going on. Its probably an install issue. Ill look into it a little more. Thanks for the rapid response!

**Comment by ercius** [comment]
Also, I am using a reader with FileVersion.VERSION4. 

Here is the full error. I noticed I did not copy the top. Its lines 287-289 where the issue is:
```
In [12]: stim.electron_count(r,np.zeros((576,576)),scan_dimensions=(513,512),background_threshold_n_sigma=4,xray_threshold_n_sigma=10)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-5592721e3cf3> in <module>
----> 1 stim.electron_count(r,np.zeros((576,576)),scan_dimensions=(513,512),background_threshold_n_sigma=4,xray_threshold_n_sigma=10)

~/temp_pae/stempy-dev/source/stempy/python/stempy/image/__init__.py in electron_count(reader, darkreference, number_of_samples, background_threshold_n_sigma, xray_threshold_n_sigma, threshold_num_blocks, scan_dimensions, verbose)
    287     data = _image.electron_count(reader.begin(), reader.end(),
    288                                  darkreference, background_threshold,
--> 289                                  xray_threshold, scan_dimensions)
    290
    291     electron_counted_data = namedtuple('ElectronCountedData',

TypeError: electron_count(): incompatible function arguments. The following argument types are supported:
    1. (arg0: stempy._io._reader_iterator, arg1: stempy._io._reader_iterator, arg2: stempy._image._image_double, arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    2. (arg0: stempy._io._sector_reader_iterator, arg1: stempy._io._sector_reader_iterator, arg2: stempy._image._image_double, arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    3. (arg0: stempy._io._pyreader_iterator, arg1: stempy._io._pyreader_iterator, arg2: stempy._image._image_double, arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    4. (arg0: stempy._io._reader_iterator, arg1: stempy._io._reader_iterator, arg2: numpy.ndarray[float64], arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    5. (arg0: stempy._io._sector_reader_iterator, arg1: stempy._io._sector_reader_iterator, arg2: numpy.ndarray[float64], arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    6. (arg0: stempy._io._pyreader_iterator, arg1: stempy._io._pyreader_iterator, arg2: numpy.ndarray[float64], arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data

Invoked with: <stempy._io._sector_reader_iterator object at 0x7f6365b28228>, <stempy._io._sector_reader_iterator object at 0x7f6365b28490>, array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]]), 0.7166298412237315, 93.54198823723709, (513, 512)
```

**Comment by cjh1** [comment]
@ercius Its definitly a python code miss match. 287 does point to that line on master. What does `git log` give you?

**Comment by ercius** [comment]
I just did another git pull and I got some new code. Ill test that.

**Comment by ercius** [comment]
Still no good. But now the error points to 290:
```
In [5]: stim.electron_count(r,np.zeros((576,576)),scan_dimensions=(513,512),background_threshold_n_sigma=4,xray_threshold_n_sigma=10,verbose=True)
****Statistics for calculating electron thresholds****
number of samples: 40
min sample: 0
max sample: 122
mean: 5.072524836033951
variance: 78.36149349359692
std dev: 8.852202748107215
number of bins: 122
x-ray threshold n sigma: 10.0
background threshold n sigma: 4.0
optimized mean: -0.26682873074717994
optimized std dev: 0.2459024312372925
background threshold: 0.71678099420199
xray threshold: 93.5945523171061
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-a84279aa3645> in <module>
----> 1 stim.electron_count(r,np.zeros((576,576)),scan_dimensions=(513,512),background_threshold_n_sigma=4,xray_threshold_n_sigma=10,verbose=True)

~/temp_pae/stempy-dev/source/stempy/python/stempy/image/__init__.py in electron_count(reader, darkreference, number_of_samples, background_threshold_n_sigma, xray_threshold_n_sigma, threshold_num_blocks, scan_dimensions, verbose)
    288     data = _image.electron_count(reader.begin(), reader.end(),
    289                                  darkreference, background_threshold,
--> 290                                  xray_threshold, scan_dimensions)
    291
    292     electron_counted_data = namedtuple('ElectronCountedData',

TypeError: electron_count(): incompatible function arguments. The following argument types are supported:
    1. (arg0: stempy::BlockIterator<stempy::StreamReader>, arg1: stempy::BlockIterator<stempy::StreamReader>, arg2: stempy._image._image_double, arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    2. (arg0: stempy::BlockIterator<stempy::SectorStreamReader>, arg1: stempy::BlockIterator<stempy::SectorStreamReader>, arg2: stempy._image._image_double, arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    3. (arg0: stempy::PyReader::iterator, arg1: stempy::PyReader::iterator, arg2: stempy._image._image_double, arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    4. (arg0: stempy::BlockIterator<stempy::StreamReader>, arg1: stempy::BlockIterator<stempy::StreamReader>, arg2: numpy.ndarray[float64], arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    5. (arg0: stempy::BlockIterator<stempy::SectorStreamReader>, arg1: stempy::BlockIterator<stempy::SectorStreamReader>, arg2: numpy.ndarray[float64], arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data
    6. (arg0: stempy::PyReader::iterator, arg1: stempy::PyReader::iterator, arg2: numpy.ndarray[float64], arg3: float, arg4: float, arg5: int, arg6: int) -> stempy._image._electron_counted_data

Invoked with: <stempy._io._sector_reader_iterator object at 0x7f66f5526308>, <stempy._io._sector_reader_iterator object at 0x7f66f55260d8>, array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]]), 0.71678099420199, 93.5945523171061, (513, 512)```

Here is the git log output:
```
commit ff328264cdbf767c0411d1f2085a9536d03e783a (HEAD -> master, origin/master, origin/HEAD)
Merge: 4e64bc4 3da759e
Author: Patrick Avery <patrick.avery@kitware.com>
Date:   Fri Jan 31 14:40:37 2020 -0500

    Merge pull request #115 from OpenChemistry/named-tuple-descriptions

    Add descriptions for our named tuples

commit 3da759e7b0aa54b41583777a2b3bd5024a8d340d (origin/named-tuple-descriptions)
Author: Patrick Avery <patrick.avery@kitware.com>
Date:   Fri Jan 31 09:35:09 2020 -0500

    Add descriptions for our named tuples

    This adds descriptions for our named tuples in the return type
    descriptions of the functions that return them.

    Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

commit 4e64bc41b506efbd1c3e4c04ec887067a3acb0f3
Merge: ee27cc6 82702ba
Author: Patrick Avery <patrick.avery@kitware.com>
Date:   Thu Jan 30 15:21:31 2020 -0500

    Merge pull request #114 from OpenChemistry/read-the-docs

    Initial setup for Read the Docs

commit 82702ba13afc5004f62972c60e9880ff5d6dce74
Author: Patrick Avery <patrick.avery@kitware.com>
Date:   Thu Jan 30 15:16:31 2020 -0500

    Use Sphinx Theme for Read the Docs

    Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

commit 50a6e2ae09d1792bfc2d5dd9186f9dc2e894bda6
Author: Patrick Avery <patrick.avery@kitware.com>
Date:   Wed Jan 29 14:50:25 2020 -0500

    Initial setup for Read the Docs

    This adds a .readthedocs.yml file, and it makes some changes to
    the docs code so that Read the Docs will work.

    The files automatically generated by `sphinx-apidoc` were committed,
    because Read the Docs does not run `sphinx-apidoc`.

    Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

commit ee27cc61c8d436c913505a12e1fbd76848c898aa
Merge: a9e8b5a f47dfa5
Author: Patrick Avery <patrick.avery@kitware.com>
Date:   Thu Jan 30 14:32:20 2020 -0500

    Merge pull request #113 from OpenChemistry/sphinx

    Mock out pybind11 generate classes

commit f47dfa519f4251ba7a7ecf1c76078b3333a3e225
Author: Chris Harris <chris.harris@kitware.com>
Date:   Wed Jan 29 22:06:20 2020 -0500

    Mock out pybind11 generate classes

    We are doing this manually to avoid problems with nested classes.

commit a9e8b5acee92377d649df9f096ab8dbef116abf7
Merge: fca41ce bf9334a
Author: Patrick Avery <patrick.avery@kitware.com>
Date:   Wed Jan 29 15:05:49 2020 -0500

    Merge pull request #112 from OpenChemistry/fix-warning

    Fix order of initialization warning

commit bf9334a8dd76065dd6f4581ca16c1ebd8b3d58b2
Author: Patrick Avery <patrick.avery@kitware.com>
Date:   Wed Jan 29 14:13:12 2020 -0500

    Fix order of initialization warning

    Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

commit fca41cef15f746c3655a2b762a43535d36581bac
```

**Comment by cjh1** [comment]
Ok, Lets revert to our know working state:

`git checkout 73bd78c13b005dcca6f4fc1df4786c857435cd81`

**Comment by cjh1** [comment]
You also copied over the new `.so`'s? After you built?

**Comment by ercius** [comment]
I went deeper to use _image.electron_count() and that works:
```python
from stempy import _image
import stempy.io as stio
r = stio.reader('/mnt/nvmedata1/temp/data_scan205_module3to4_dst1_file0.data',stio.FileVersion.VERSION4) 
_image.electron_count(r.begin(),r.end(),np.zeros((576,576)),28.0,100.0,513,512)
```
This worked. So there is just an issue with type checking somewhere in the python code.

**Comment by ercius** [comment]
> You also copied over the new .so's? After you built?
Ill double check this and make sure its correct.

Then Ill pull the old state. That will at least get me back to a working state.

**Comment by ercius** [comment]
ok...it looks good now. I think its fixed.

I had to really start from scratch. The initial pull did not update fully to the newest stempy on master. I did another pull and that eventually got me fully updated. 

Then I had to re-run the build and cp the *.so properly. Then I had to pip install -U -e to get the python working.

So, I think we are all good. Not sure why `git pull` did not work the first time around. Thanks for all your help!
